include/boost/beast/core/coroutine.hpp

```diff
@@ -0,0 +1,30 @@
+//
+// Copyright (c) 2016-2017 Vinnie Falco (vinnie dot falco at gmail dot com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/boostorg/beast
+//
+
+#ifndef BOOST_BEAST_COROUTINE_HPP
+#define BOOST_BEAST_COROUTINE_HPP
+
+#include <boost/asio/coroutine.hpp>
+
+namespace boost {
+namespace beast {
+
+inline bool is_continuation(asio::coroutine& coroutine) noexcept
+{
+    asio::detail::coroutine_ref ref{coroutine};
+    // We need to reassign the current coroutine state, because
+    // coroutine_ref marks the coroutine as complete if it is not modified.
+    auto state = (ref = static_cast<int>(ref));
+    return state > 0;
```

Comment:
Oh damn, I see what you did here, this is very smart!

*BUT*..... we can't be peeking into the `asio::detail::` namespace. And I don't think this should be a public interface.

An alternative is to just copy all the Asio stackless coroutine code and then it becomes ours to do with as we please.

Reply:
If you DO add a public header file you need to add it to the "convenience header", in this case `<boost/beast/core.hpp>`.

Reply:
While I like the idea, we have a problem. I have another branch waiting which updates Beast to use the new version of Boost.Asio which is styled after the Networking-TS. See: https://github.com/vinniefalco/beast/commits/net-ts

The new version of Boost.Asio removes the `asio_is_continuation` hook and replaces it with free functions `post` and `defer` which perform similar behaviors. But of course it completely changes the program logic. So I don't think its a good idea for us to change the `is_continuation` code otherwise we will get pretty bad merge conflicts. This change will have to wait until after the new branch is merged.

That said, I like the technique of leveraging the `coroutine` state to know if we are already in the proper context. I like it so much that I think it should be a member function of `asio::coroutine`. I think you should submit it as a pull request.